llama_index/callbacks/open_inference_callback.py

```diff
@@ -0,0 +1,315 @@
+    """
+    try:
+        package = importlib.import_module(package_name)
+    except ImportError:
+        raise ImportError(f"The {package_name} package must be installed.")
+    return package
+
+
+class DataBuffer(Generic[BaseDataType]):
+    def __init__(self, max_size_in_bytes: Optional[int] = None) -> None:
+        """The data buffer for OpenInference data.
+
+        Args:
+            max_size_in_bytes (Optional[int], optional): The maximum size of the
+                buffer in bytes. Once this size is reached, the oldest data will
+                be removed until the size of the buffer is less than the maximum
+                size.
+        """
+        self._id_to_data_map: OrderedDict[str, BaseData] = OrderedDict()
+        self._max_size_in_bytes = max_size_in_bytes
+        self._size_in_bytes = 0
+
+    def __len__(self) -> int:
+        return len(self._id_to_data_map)
+
+    def add(self, data: BaseData) -> None:
+        """Add data to the buffer. If the buffer is full, the oldest data will
+        be removed until the size of the buffer is less than the maximum
+        configured size.
+
+        Args:
+            data (BaseData): Data to add to the buffer.
+        """
+        if data.id in self._id_to_data_map:
+            self._size_in_bytes -= sys.getsizeof(self._id_to_data_map.pop(data.id))
+        self._id_to_data_map[data.id] = data
+        self._size_in_bytes += sys.getsizeof(data)
+
+        if self._max_size_in_bytes is not None:
+            while self._size_in_bytes > self._max_size_in_bytes:
+                _, data = self._id_to_data_map.popitem(last=False)
+                self._size_in_bytes -= sys.getsizeof(data)
+
+    def clear(self) -> None:
+        """Clears the buffer."""
+        self._id_to_data_map.clear()
+        self._size_in_bytes = 0
+
+    @property
+    def dataframe(self) -> "DataFrame":
+        """Returns the data buffer as a pandas dataframe.
+
+        Returns:
+            DataFrame: The buffer dataframe.
+        """
+        pandas = _import_package("pandas")
+        return pandas.DataFrame(
+            [data.to_dict() for data in self._id_to_data_map.values()]
+        )
+
+    @property
+    def size_in_bytes(self) -> int:
+        """Returns the size of the buffer in bytes.
+
+        Returns:
+            int: The size of the buffer in bytes.
+        """
+        return self._size_in_bytes
+
+
+class OpenInferenceCallbackHandler(BaseCallbackHandler):
+    def __init__(
+        self,
+        max_size_in_bytes: Optional[int] = None,
+        logger: Optional[Callable[[DataBuffer[QueryData]], None]] = None,
+    ) -> None:
+        """Initializer for the OpenInferenceCallbackHandler.
+
+        Args:
+            max_size_in_bytes (Optional[int], optional): The buffer size
+                threshold. If this argument is set to None, the buffer will not
+                be limited in size.
+
+            logger (Optional[Callable[[OpenInferenceDataBuffer[QueryData]],
```

Comment:
```suggestion
            on_append_callback (Optional[Callable[[OpenInferenceDataBuffer[QueryData]],
```

Reply:
Changed the recommendation from creating a custom logger to creating a sub-class of the `OpenInferenceCallbackHandler` to handle logging logic. About to update the tutorial notebook, let me know what you think.

Reply:
I rather not have to sub-class things. I liked the way you had it.

Reply:
I think I agree. Addressed in [a57274c](https://github.com/jerryjliu/llama_index/pull/6998/commits/a57274c0633dbaf1019685241d5312fa475550aa).